src/hardware/gus.cpp

```diff
@@ -715,17 +719,37 @@ void Gus::CheckVoiceIrq()
 	}
 }
 
-uint32_t Gus::Dma8Addr() noexcept
+// Returns a 24-bit offset into the GUS's memory space holding the next
+// DMA sample that will be read or written to via DMA. This offset
+// is derived from the 16-bit DMA address register.
+uint32_t Gus::GetDmaOffset() noexcept
 {
-	return static_cast<uint32_t>(dma_addr << 4);
+	uint32_t adjusted;
+	if(IsDmaXfer16Bit()) {
+		const auto upper = dma_addr & 0b1100'0000'0000'0000;
+		const auto lower = dma_addr & 0b0001'1111'1111'1111;
+		adjusted = static_cast<uint32_t>(upper | (lower << 1));
+	}
+	else {
+		adjusted = dma_addr;
+	}
+	return check_cast<uint32_t>(adjusted << 4) + dma_addr_nibble;
 }
 
-uint32_t Gus::Dma16Addr() noexcept
+// Update the current 16-bit DMA position from the the given 24-bit RAM offset 
+void Gus::UpdateDmaAddr(uint32_t offset) noexcept
 {
-	const auto lower = dma_addr & 0b0001'1111'1111'1111;
-	const auto upper = dma_addr & 0b1100'0000'0000'0000;
-	const auto combined = (lower << 1) | upper;
-	return static_cast<uint32_t>(combined << 4);
+	uint32_t adjusted;
+	if (IsDmaXfer16Bit()) {
+		const auto upper = offset & 0b1100'0000'0000'0000'0000;
+		const auto lower = offset & 0b0011'1111'1111'1111'1110;
+		adjusted = upper | (lower >> 1);
+	}
+	else {
+		adjusted = offset;
+	}
+	dma_addr = check_cast<uint16_t>(adjusted >> 4); // pack it into the 16-bit register
```

Comment:
This line is causing Quake to assert with GUS enabled. If `adjusted` is a 24-bit value, `adjusted >> 4` will give a 20-bit value, causing the `check_cast` to fail. Do we want this to be either a `static_cast` or change it to e.g.
```cpp
	dma_addr = check_cast<uint16_t>(adjusted >> 4 & 0xffff);
```
?

Reply:
Quake output:
```log
022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fa00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fb00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fc00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fd00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = fe00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = ff00
2022-06-18 21:50:09.595 | GUS: DMA address adjusted >> 4 = 10000
Assertion failed: (static_cast<next_int_t<check_t>>(in) <= static_cast<next_int_t<cast_t>>(std::numeric_limits<cast_t>::max())), function check_cast, file support.h, line 208.
```

Reply:
Thank you, @kklobe! Right; with exactly 1 MB of ram, that's 20 bits of addressable space  (2 ** 20 = 1048576), and given the first 4 bits are shifted off the end, that leaves the top 16bit in the 20 bit range. :+1:  Applied and regression tested.